README.md
[![ci-cd](https://github.com/UBC-MDS/sanityze/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/sanityze/actions/workflows/ci-cd.yml) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![Documentation Status](https://readthedocs.org/projects/sanityze/badge/?version=latest)](https://sanityze.readthedocs.io/en/latest/?badge=latest) ![PyPI](https://img.shields.io/pypi/v/sanityze) [![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active) [![Python 3.9+](https://img.shields.io/badge/python-3.9-blue.svg)](https://www.python.org/downloads/release/python-390/)

# sanityze

![](logo.png)

Data scientists often need to remove or redact Personal Identifiable Information (PII) from their data. This package provides utilities to spot and redact PII from Pandas data frames.

PII can be used to uniquely identify a person. This includes names, addresses, credit card numbers, phone numbers, email addresses, and social security numbers, and therefore regulatory bodies such as the European Union's General Data Protection Regulation (GDPR) and the California Consumer Privacy Act (CCPA) require that PII be removed or redacted from data sets before they are shared an further processed.

## Contributors and Maintainers

- [Tony Zoght](https://github.com/tzoght)
- [Caesar Wong](https://github.com/caesarw0)
- [Jonah Hamilton](https://github.com/xXJohamXx)

## Why `sanityze` ?

Because it's a fun name and it's a play on the word "sanitize" which is what we are doing to the data.

## Similar packages in Python

The closet Python package in functionality to sanityze is  [scrubadub](https://scrubadub.readthedocs.io/en/stable/) which is a package for finding and removing PII from text. The package is not designed to work with Pandas data frames, or other data structures, and we believe that our package will be more useful to data scientists, as we add more spotters (mechanisms for finding PII), support for more data structures, and provide mechanisms for users to define their own spotters.

## Quick Start

To get started with `sanityze`, install it using `pip`:

```bash
pip install sanityze
```

And visit the [documentation](https://ubc-mds.github.io/sanityze/) for more information and examples.

## Features and Usage

Conceptually, `sanityze` is a package that provides a way to remove PII from Pandas data frames. The package provides a number of default spotters, which can be used to identify PII in the data and redact them.

The main entry point to the package is the `Cleanser` class. The `Cleanser` class is used to add `Spotter`s to the cleanser, which will be used to identify PII in the data. The cleanser can then be used to cleanse the data, and redact the PII from the given data frame (all future data structures that will be suppportd by the package, in the future).

The package comes with a number of default spotters, as subclassess of `Spotter`:

1. `CreditCardSpotter` - identifies credit card numbers
2. `EmailSpotter` - identifies email addresses

Spotters can be added to it using the `add_spotter()` method. The cleanser can then be used to cleanse data using the `cleanse()` method which takes a Pandas data frame and returns a Pandas data frame with PII redacted.

The redaction options provided by `sanityze`` are:

1. Redact using a fixed string - The string in this case is the ID of the spotter. For example, if the spotter is an instance of `CreditCardSpotter`, the string will be `{{CREDITCARD}}`, or `{{EMAILADDRS}}` for an instance of `EmailSpotter`.
2. Redact using a hash of the input - The hash is computed using the `hashlib` package, and the hash function is `md5`. For example, if the spotter is an instance of `CreditCardSpotter`, the string will be `{{6a8b8c6c8c62bc939a11f36089ac75dd}}`, if the input is contains a PII `1234-5678-9012-3456`.

## Classes and Functions

1. `Cleanser`: the main class of the package. It is used to add spotters to it, and then cleanse data using the spotters.
   1. `add_spotter()`: adds a spotter to the cleanser
   2. `remove_spotter()`: removes a spotter from the cleanser
   3. `clean()`: cleanses the data in the given data frame, and returns a new data frame with PII redacted
2. `EmailSpotter`: a spotter that identifies email addresses
   1. `getUID()`: returns the unique ID of the spotter
   2. `process()`: performs the PII matching and redaction
3. `CreditCardSpotter`: a spotter that identifies credit card numbers
   1. `getUID()`: returns the unique ID of the spotter
   2. `process()`: performs the PII matching and redaction

> You can checkout detailed API Documentations [here](https://ubc-mds.github.io/sanityze/).

Below is a simple quick start example:

```python
import pandas as pd
from sanityze import Cleanser, EmailSpotter

# Create a cleanser, and don't add the default spotters
cleanser = Cleanser(include_default_spotters=False)
cleaner.add_spotter(from sanityze import Cleanser, EmailSpotter())
cleaned_df = cleanser.clean(df)
```

## High-level Design

To better understand the design of the package, we have provided a high-level design document, which will be kept up to date as the package evolves. The document can be found [here](HighLevelDesign.md).

## Contributing

Interested in contributing? Check out the [contributing guidelines](CONTRIBUTING.md). Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`sanityze` was created by Caesar Wong, Jonah Hamilton and Tony Zoght. It is licensed under the terms of the [MIT license](LICENSE).

## Credits

`sanityze` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).

## Quick Links

- [PyPI](https://pypi.org/project/sanityze/)
- [Read the Docs](https://sanityze.readthedocs.io/en/latest/?badge=latest)
- [Documentation on GH](https://ubc-mds.github.io/sanityze/)
- [Kanban Board](https://github.com/orgs/UBC-MDS/projects/15)
- [Issues](https://github.com/UBC-MDS/sanityze/issues)
- [High Level Design](HighLevelDesign.md)
- [Contributing Guidelines](CONTRIBUTING.md)
- [Code of Conduct](CODE_OF_CONDUCT.md)
- [License](LICENSE)


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.2 (2023-02-04)


## v1.0.1 (2023-02-04)


## v1.0.0 (2023-02-04)
### Feature
* Major release bump ([`7e4ea82`](https://github.com/UBC-MDS/sanityze/commit/7e4ea82d1cbd0b382e0d155e52f03b3a58409fac))

### Breaking
* version 1.0.0 ([`7e4ea82`](https://github.com/UBC-MDS/sanityze/commit/7e4ea82d1cbd0b382e0d155e52f03b3a58409fac))

## v0.1.3 (2023-01-27)
### Documentation
* Add PyPI badge ([`f1a7593`](https://github.com/UBC-MDS/sanityze/commit/f1a75938f11e0d3425f847e28bfee338b39f31f5))

## v0.1.2 (2023-01-27)
### Documentation
* Adding links to PyPI and ReadTheDocs ([`a402d4f`](https://github.com/UBC-MDS/sanityze/commit/a402d4f7372a40f218e9b842f0b02432d50d9c86))
* Updated readme to point to PyPI ([`98f29f2`](https://github.com/UBC-MDS/sanityze/commit/98f29f237a900d2eb1df23bd03bdc5f560186647))

## v0.1.1 (2023-01-27)


## v0.1.0 (2023-01-27)
### Feature
* Adding support for documentation ([`b8cba13`](https://github.com/UBC-MDS/sanityze/commit/b8cba1300a3c57a9ab0dc76ba8470ddf75921033))

## v0.1.0 (11/01/2023)

- First release of `sanityze`!

tests/test_cleanser.py
from sanityze.cleanser import *
from sanityze.spotters import *
import pytest 

input_data = {'product_name': ['my email is foo@gaga.com', 
                            'Okay', 
                            'my number is 4556129404313766 CCV'],
        'price': [1200, 150, 300]
    }
input_df = pd.DataFrame(input_data)
expected_data = {'product_name': ['my email is EMAILADDRS', 
                            'Okay', 
                            'my number is CREDITCARD CCV'],
        'price': [1200, 150, 300]
    }
expected_df = pd.DataFrame(expected_data)

# check that the Cleanser class is initialized property
def test_cleanser_init_default():
    c = Cleanser()
    assert(len(c.chain) > 0,"Cleanser should have at least one spotter in the chain")

# Init with no spotters
def test_cleanser_init_no_spotters():
    c = Cleanser(include_default_spotters=True)
    assert(len(c.chain) == 0,"Cleanser should have no spotters in the chain when include_default_spotters is False")
  
# Init with no spotters, add/remove later  (EmailSpotter)
def test_cleanser_add_remove_e():
    c = Cleanser(include_default_spotters=False)
    s = EmailSpotter("EMAILS",False)
    c.add_spotter(s)
    assert(len(c.chain) == 1,"Cleanser should have one spotter in the chain")
    assert(c.chain[0] == s,"Cleanser should have the EmailSpotter in the chain")
    c.remove_spotter("EMAILS")
    assert(len(c.chain) == 0,"Cleanser should have no spotters in the chain")

# Init with no spotters, add/remove later  (CreditCardSpotter)
def test_cleanser_add_remove_c():
    c = Cleanser(include_default_spotters=False)
    s = CreditCardSpotter("CREDITCARDS",False)
    c.add_spotter(s)
    assert(len(c.chain) == 1,"Cleanser should have one spotter in the chain")
    assert(c.chain[0] == s,"Cleanser should have the CreditCardSpotter in the chain")
    c.remove_spotter("CREDITCARDS")
    assert(len(c.chain) == 0,"Cleanser should have no spotters in the chain")
    
# Init with no spotters, add/remove all spotters later  (EmailSpotter, CreditCardSpotter)
def test_cleanser_add_remove_all():
    c = Cleanser(include_default_spotters=False)
    s1 = EmailSpotter("EMAILS",False)
    s2 = CreditCardSpotter("CREDITCARDS",False)
    c.add_spotter(s1)
    assert(len(c.chain) == 1,"Cleanser should have one spotter in the chain")
    c.add_spotter(s2)
    assert(len(c.chain) == 2,"Cleanser should have two spotters in the chain")
    c.remove_spotter("EMAILS")
    assert(len(c.chain) == 1,"Cleanser should one spotter in the chain")
    c.remove_spotter('CREDITCARDS')
    assert(len(c.chain) == 1,"Cleanser should no spotters in the chain")
    
# test adding duplicate spotters
def test_add_duplicate_spotters():
    c = Cleanser(include_default_spotters=False)
    s1 = EmailSpotter("EMAILS",False)
    s2 = EmailSpotter("EMAILS",False)
    a1 = c.add_spotter(s1)
    a2 = c.add_spotter(s2)
    assert(a1 == True,"Cleanser should have added the first spotter")
    assert(a2 == False,"Cleanser should not have added the second spotter")
    assert(len(c.chain) == 0,"Cleanser should have not spotters in the chain")

# test basic clean   
def test_basic_clean():
    c = Cleanser(include_default_spotters=True)
    output_df = c.clean(input_df)
    assert(output_df.equals(expected_df),"Cleanser should have cleaned the data properly")
    
# test adding None spotter
def test_add_none_spotter():
    with pytest.raises(ValueError):
        c = Cleanser(include_default_spotters=False)
        a = c.add_spotter(None)
        
# test removing None spotter
def test_remoe_none_spotter():
    with pytest.raises(ValueError):
        c = Cleanser(include_default_spotters=False)
        a = c.remove_spotter(None)
        
# test adding duplicate spotters
def test_add_duplicate_spotters():
    c = Cleanser(include_default_spotters=False)
    s1 = EmailSpotter("EMAILS",False)
    a1 = c.add_spotter(s1)
    a2 = c.add_spotter(s1)
    assert(a1 == True,"Cleanser should have added the first spotter")
    assert(a2 == False,"Cleanser should not have added the second spotter")
    assert(len(c.chain) == 0,"Cleanser should have not spotters in the chain")
    
# test passing a non-DataFrame to clean
def test_clean_with_non_dataframe():
    c = Cleanser(include_default_spotters=True)
    with pytest.raises(TypeError):
        output_df = c.clean("foo")
        
# utility function to test the log
def test_print_log_1():
    c = Cleanser(include_default_spotters=True)
    c._log("foo",True)
    assert(True,"Cleanser should have printed the log properly")
    
# utility function to test the log
def test_print_log_2():
    c = Cleanser(include_default_spotters=True)
    c._log("foo",False)
    assert(True,"Cleanser should have not printed the log properly")

tests/test_sanityze.py
from sanityze.cleanser import *
from sanityze.spotters import *

def test_hello_world():
    assert True == True



tests/test_dummy_data.py
from sanityze.cleanser import *
from sanityze.spotters import *
import pandas as pd

def test_cleanser_with_dummy_data():

    # create Cleanser with default spotter
    c = Cleanser(include_default_spotters=True)

    assert len(c.chain) == 2,"Cleanser should have two spotter in the chain by default"

    # reading dummy data with pii
    df_with_pii = pd.read_csv('tests/data_with_pii.csv')
    assert df_with_pii.shape == (20, 8), "Dummy Data with PII should have shape (20, 8)"
    # reading cleaned dummy data without hash
    df_with_pii_cleaned_non_hash = pd.read_csv('tests/data_with_pii_cleaned_non_hash.csv')
    assert c.clean(df_with_pii).equals(df_with_pii_cleaned_non_hash), "PII should be cleaned with a fixed string"


    # reading dummy data without pii
    df_without_pii = pd.read_csv('tests/data_without_pii.csv')
    df_without_pii_copy = df_without_pii.copy()
    assert df_without_pii.shape == (20, 5), "Dummy Data without PII should have shape (20, 8)"

    assert c.clean(df_without_pii).equals(df_without_pii_copy), "Data without PII should be the same after c.clean()"
    


    # create Cleanser with hash spotter
    c = Cleanser(include_default_spotters=False)
    s1 = EmailSpotter("EMAILS",True)
    s2 = CreditCardSpotter("CREDITCARDS",True)
    c.add_spotter(s1)
    c.add_spotter(s2)

    assert len(c.chain) == 2,"Cleanser should have two spotter in the chain by default"

    # reading cleaned dummy data with hash
    df_with_pii_cleaned_hash = pd.read_csv('tests/data_with_pii_cleaned_hash.csv')
    assert c.clean(df_with_pii).equals(df_with_pii_cleaned_hash), "PII should be cleaned with hash"

    assert c.clean(df_without_pii).equals(df_without_pii_copy), "Data without PII should be the same after c.clean()"


src/sanityze/cleanser.py
import pandas as pd
from sanityze.spotters import * 

class Cleanser:
    """
    The main class for the sanityze package. It's purpose is to clean the data frame
    before it's consumed by the training or prediction pipeline.
    
    Parameters
    ----------
    include_default_spotters : bool, optional
        If True, the default spotters will be added to the Cleanser. The default is True.
    hash_spotted : bool, optional
        If True, the spotters will hash the values within the columns they spot. 
        The default is False.
    
    """
    def __init__(self, include_default_spotters=True, hash_spotted = False):
        if (include_default_spotters):
            self.chain = [EmailSpotter("DEFAULTEMAILS",hash_spotted),CreditCardSpotter("DEFAULCCS",hash_spotted)]
        else:
            self.chain = []
    
    def add_spotter(self, spotter) -> bool:
        """
        Add a specific spotter to the Cleanser
        
        Parameters
        ----------
        spotter : Spotter
            A subclass of Spotter to add to the Cleanser. Note that spotters are added
            at the end of the list. Adding the same spotter will return False

        Returns
        -------
        True if the spotter was added, False if it was not added.   

        Examples
        --------
        >>> c = Cleanser(include_default_spotters=False)
        >>> s1 = EmailSpotter("EMAILS",True)
        >>> c.add_spotter(s1)

        """
        if (spotter is None):
            raise ValueError("spotter cannot be None in Cleanser.add_spotter()")
        if (spotter in self.chain):
            return False
        for s in self.chain:
            if (s.getSpotterUID() == spotter.getSpotterUID()):
                return False
        self.chain.append(spotter)
    
    def remove_spotter(self, spotter_id) -> bool:
        """
        Remove a specific spotter from the Cleanser using the spotter's id
        
        Parameters
        ----------
        spotter_id : str
            The id of the spotter to remove
        verbose: bool, optional
            If True, the spotter will print out debug information. The default is False.
        
        Returns
        -------
        True if the spotter was removed, False if it was not removed.

        Examples
        --------
        >>> c = Cleanser(include_default_spotters=False)
        >>> s1 = EmailSpotter("EMAILADDRS",True)
        >>> c.remove_spotter("EMAILADDRS")

        """
        if (spotter_id is None):
            raise ValueError("spotter_id cannot be None in Cleanser.remove_spotter()")
        for s in self.chain:
            if (s.getSpotterUID() == spotter_id):
                self.chain.remove(s)
                return True
        return False
    
    def _log(self, message: str, verbose: bool) -> None:
        """
        Internal utility function to log messages to the console
        
        Parameters
        ----------
        verbose: bool
            The verbosity of the log
        message : str
            The message to log
        
        Returns
        -------
        None

        Examples
        --------
        (called by clean())

        """
        if (verbose):
            print(f"- {message}")

    def clean(self, df: pd.DataFrame, verbose=False) -> pd.DataFrame:
        """
        Sanitizes the data frame using the spotters added to the Cleanser
        
        Parameters
        ----------
        df : pd.DataFrame
            The data frame to sanitize
            
        Returns
        -------
        The sanitized data frame 
        
        Examples
        --------
        >>> df = pd.DataFrame(data = {'product_name': ['laptop', 'printer foo@gaga.com', 'tablet', 'desk 5555 5555 5555 4444', 'chair'],
                                    'price': [1200, 150, 300, 450, 200]})
        >>> c = Cleanser()
        >>> c.clean(df, verbose=False)
            product_name	price
        0	laptop	1200
        1	printer EMAILADDRS	150
        2	tablet	300
        3	desk 5555 5555 5555 4444	450
        4	chair	200
        
        """
        if (df is None):
            raise ValueError("df cannot be None in clean")
        if not isinstance(df,pd.DataFrame):
            raise TypeError("df must be a pandas DataFrame in clean")
        # we only operate on a copy of the data frame, leaving 
        # the original data frame intact
        df_copy = df.copy() 
        # iterate thru the data frame cells
        row_len, col_len = df.shape
        for i in range(row_len):
            for j in range(col_len):
                cell = df.iat[i,j]
                # iterate thru the spotters and redact content
                # if the cell is of type string
                if isinstance(cell, str):
                    for spotter in self.chain:
                        self._log(f"{spotter.getSpotterUID()}: Processing cell {cell} ", verbose)
                        cell = spotter.process(cell)
                        self._log(f"{spotter.getSpotterUID()}: Processed cell {cell} ", verbose)
                # update the cell in the copy of the data frame
                df_copy.iat[i,j] = cell
        return df_copy

src/sanityze/sanityze.py
import pandas as pd

class Cleanser:
    """
    The main class for the sanityze package. It's purpose is to clean the data frame
    before it's consumed by the training or prediction pipeline.
    
    Parameters
    ----------
    include_default_spotters : bool, optional
        If True, the default spotters will be added to the Cleanser. The default is True.
    hash_spotted : bool, optional
        If True, the spotters will hash the values within the columns they spot. 
        The default is False.
    
    """
    def __init__(self, include_default_spotters=True, hash_spotted = False):
        pass
    
    """
    Add a specific spotter to the Cleanser
    
    Parameters
    ----------
    spotter : Spotter
        A subclass of Spotter to add to the Cleanser. Note that spotters are added
        at the end of the list. Adding the same spotter will return False

    Returns
    -------
    True if the spotter was added, False if it was not added.    
    """
    def add_spotter(self, spotter) -> bool:
        pass
    
    """
    Remove a specific spotter from the Cleanser using the spotter's id
    
    Parameters
    ----------
    spotter_id : str
        The id of the spotter to remove
    
    Returns
    -------
    True if the spotter was removed, False if it was not removed.
    """
    def remove_spotter(self, spotter_id) -> bool:
        pass
    
    """
    Sanitizes the data frame using the spotters added to the Cleanser
    
    Parameters
    ----------
    df : pd.DataFrame
        The data frame to sanitize
        
    Returns
    -------
    The sanitized data frame 
    
    """
    def clean(self, df: pd.DataFrame) -> pd.DataFrame:
        pass

src/sanityze/spotters.py
import re
import hashlib

class Spotter():
    """
    The Spotter interface to be implemented

    Attributes
    ----------
    uid : str
        uid of the spotter
    hashSpotted : bool, optional
        False by default, whether to hash or replace the spotted sensitive information

    Methods
    -------
    getSpotterUID()
        return the Spotter uid
    
    isHashSpotted()
        return whether the hashSpotted is True or False
    
    process(text)
        process the text depending on the hashSpotted value, if it is hash, replace it with hash
        otherwise, replace it with some default value

    Examples
    --------
    Spotter should be initialized in a subclass level, therefore, skipping examples in the parent class
    >>> 

    """
    # please add the following line in the subclass
    # spotter_uid = "<uid of the spotter>"
    def __init__(self, uid: str, hashSpotted=False):
        self.uid = uid
        self.hashSpotted = hashSpotted

    def getSpotterUID(self) -> str:
        """Getting the spotter uid

        Returns
        -------
        self.uid : str
            the spotter uid

        Examples
        --------
        >>> sub_spotter.getSpotterUID()
        "<sub class spotter UID>"
        """
        return self.uid

    def isHashSpotted(self) -> bool:
        """Getting the value of hashSpotted

        Returns
        -------
        self.hashSpotted : bool
            the Truth value of hashSpotted
        
        Examples
        --------
        >>> sub_spotter.isHashSpotted()
        TRUE
        """
        return self.hashSpotted

    def process(self, text: str) -> str:
        """Process the given text, if hashSpotted is True, replace the spotted text with hash,
        otherwise, replace the spotted text with some default values
        
        Parameters
        ----------
        text : str
            The text to be spotted & modified

        Returns
        -------
        new_text : str

        Examples
        --------
        >>> df = pd.DataFrame(data = {'product_name': ['laptop', 'printer foo@gaga.com', 'tablet', 'desk 5555 5555 5555 4444', 'chair'],
                                    'price': [1200, 150, 300, 450, 200]})
        >>> c = Cleanser()
        >>> c.clean(df, verbose=False)
            product_name	price
        0	laptop	1200
        1	printer EMAILADDRS	150
        2	tablet	300
        3	desk 5555 5555 5555 4444	450
        4	chair	200

        
        """
        # # to be implemented in the specific spotter level
        # if self.isHashSpotted():
        #     new_text = "hash"
        # else:
        #     new_text = ""
        # return new_text
        pass


class CreditCardSpotter(Spotter):
    """
    The Credit Card Spotter Subclass

    Attributes
    ----------
    uid : str
        uid of the spotter, "CREDITCARD"
    hashSpotted : bool, optional
        False by default, whether to hash or replace the spotted sensitive information

    Methods
    -------
    getSpotterUID()
        return the Spotter uid, "CREDITCARD"
    
    isHashSpotted()
        return whether the hashSpotted is True or False
    
    process(text)
        process the text depending on the hashSpotted value, if hashSpotted is True, replace the spotted credit card number with hash
        otherwise, replace the spotted credit card number with some default value
    
    Examples
    --------
    >>> CreditCardSpotter("CREDITCARDS",True)
    <sanityze.spotters.CreditCardSpotter object at 0x000001207F7B5880>

    
    """
    def getSpotterUID(self) -> str:
        """Getting the credit card spotter uid

        Returns
        -------
        "CREDITCARD" : str
            a fixed str value for CreditCardSpotter

        Examples
        --------
        >>> cc = CreditCardSpotter("CREDITCARDS",True)
        >>> cc.getSpotterUID()
        CREDITCARD
        
        """
        return "CREDITCARD"

    def process(self, text: str) -> str:
        """Process the given text, if hashSpotted is True, replace the spotted credit card number with hash,
        otherwise, replace the spotted credit card number with some default values
        
        Parameters
        ----------
        text : str
            The text to be spotted & modified

        Returns
        -------
        new_text : str
            the text with credit card number replaced by a hash or the default string value
        
        Examples
        --------
        >>> cc = CreditCardSpotter("CREDITCARDS", False)
        >>> cc.process("4556129404313766")
        CREDITCARD

        """
        # Regexes from:
        # http://www.regular-expressions.info/creditcard.html

        # taken from the alphagov fork of scrubadub: https://github.com/alphagov/scrubadub

        # credit card patterns to match
        cc_pattern = re.compile((
            r"(?:4[0-9]{12}(?:[0-9]{3})?"  		# Visa
            r"|(?:5[1-5][0-9]{2}"          		# MasterCard
            r"|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}"
            r"|3[47][0-9]{13}"             		# American Express
            r"|3(?:0[0-5]|[68][0-9])[0-9]{13}"   	# Diners Club
            r"|6(?:011|5[0-9]{2})[0-9]{12}"      	# Discover
            r"|(?:2131|1800|35\d{3})\d{11})"      	# JCB
         ), re.VERBOSE)

        # sets replacement value based on output of isHashSpotted()
        if self.isHashSpotted():
            replacement = hashlib.md5(text.encode()).hexdigest()
        else:
            replacement = self.getSpotterUID()
        
        # replaces cc number with replacement value
        clean_str = re.sub(cc_pattern, replacement, text)

        return(clean_str)


class EmailSpotter(Spotter):
    """
    The Email Spotter Subclass

    Attributes
    ----------
    uid : str
        uid of the spotter, "EMAILADDRS"
    hashSpotted : bool, optional
        False by default, whether to hash or replace the spotted sensitive information

    Methods
    -------
    getSpotterUID()
        return the Spotter uid, "EMAILADDRS"
    
    isHashSpotted()
        return whether the hashSpotted is True or False
    
    process(text)
        process the text depending on the hashSpotted value, if hashSpotted is True, replace the spotted email with hash
        otherwise, replace the spotted email with some default value
    """
    def getSpotterUID(self) -> str:
        """Getting the email spotter uid

        Returns
        -------
        "EMAILADDRS" : str
            a fixed str value for EmailSpotter

        Examples
        --------
        >>> ee = EmailSpotter("EMAILS", False)
        >>> ee.getSpotterUID()
        EMAILADDRS

        """
        return "EMAILADDRS"

    def process(self, text: str) -> str:
        """Process the given text, if hashSpotted is True, replace the spotted email with hash,
        otherwise, replace the spotted email with some default values
        
        Parameters
        ----------
        text : str
            The text to be spotted & modified

        Returns
        -------
        new_text : str
            the text with email replaced by a hash or the default string value

        Examples
        --------
        >>> ee = EmailSpotter("EMAILS", False)
        >>> ee.process("abcd1234@gmail.com")
        EMAILADDRS

        """
        # base preprocessing (if needed)
                
        # email regex (adapted from [https://scrubadub.readthedocs.io/en/stable/_modules/scrubadub/detectors/email.html#EmailDetector:~:text=regex%20%3D%20re,.IGNORECASE)])
        regex = re.compile((
            r"\b[a-z0-9!#$%&'*+\/=?^_`{|}~-]"             # start with this character
            r"(?:"
            r"    [\.a-z0-9!#$%&'*+\/=?^_`{|}~-]{0,62}"   # valid next characters (max length 64 chars before @)
            r"    [a-z0-9!#$%&'*+\/=?^_`{|}~-]"           # end with this character
            r")?"
            r"(?:@|\sat\s)"                               # @ or the word 'at' instead
            r"[a-z0-9]"                                   # domain starts like this
            r"(?:"
            r"    (?=[a-z0-9-]*(\.|\sdot\s))"             # A lookahead to ensure there is a dot in the domain
            r"    (?:\.|\sdot\s|[a-z0-9-]){0,251}"        # might have a '.' or the word 'dot' instead
            r"    [a-z0-9]"                               # domain has max 253 chars, ends with one of these
            r")+\b"
        ), re.VERBOSE | re.IGNORECASE)
        
        if self.isHashSpotted():            
            text = re.sub(regex, lambda x:hashlib.md5(x.group().encode()).hexdigest(), text)
            new_text = text
        else:
            new_text = re.sub(regex, self.getSpotterUID(), text)

        return new_text